### DISCUSSION TITLE
Nested ItemsControls become slower everytime item source changes

### DISCUSSION THREAD

**Discussion by butaixianran** [AvaloniaUI/Avalonia#14665] [General]
Following video shows an app developed with Avalonia, each space (list on left) has a list of groups, and each group has a title and a list of tiles.

In this video, each space has an ItemsControl for groups. And in this ItemsControl, has another ItemsControl for tiles.

As you can see, every time ItemsControl's source changes, the UI becomes slower and slower. And if this video goes on, about 1 minute later, I have to wait 8~10 seconds for UI's rendering after each modification.

https://github.com/AvaloniaUI/Avalonia/assets/5558722/5321ac15-c63b-4647-9657-5e9d884cdd62

And in this video, when draging a tile, each group will add an empty tile as placeholder. This placeholder tile, is added with `ItemControl.Items.Add()`, not even refresh the whole item sources. But UI still become slower and slower. 

I have tried to replace all StackPanel with Grid, no good at all. 

Any idea or solution for this performance issue? Nested list is a common need for most apps. 

System info:
* Avalonia version: v11~v11.09
* Platform: win 11

**Comment by thevortexcloud** [comment]
Are you using a virtualised panel for your lists? If so this sounds very similar to a problem I ran into awhile back. 

https://github.com/AvaloniaUI/Avalonia/issues/10968

However the fix should have been backported. Although the problem there was due to things being incorrectly realised.

**Comment by butaixianran** [comment]
I'm not using virtualised panel. 

In this app, ItemControl's `ItemsPanelTemplate` is using `WrapPanel`. WrapPanel is not a virtualised panel in avalonia v11, right?

For detail, `ScrollViewer` is used. `VirtualizingStackPanel` is not used. Also, I have tried replacing all `StackPanel` with `Grid`. Still got the same result.

The code for those 2 ItemControls is something like following:

Group List:
```xml
	<ScrollViewer Name="TileGroupListSV">
		<ItemsControl Name="TileGroupListIC">
			<ItemsControl.ItemsPanel>
				<ItemsPanelTemplate>
					<WrapPanel />
				</ItemsPanelTemplate>
			</ItemsControl.ItemsPanel>
			<ItemsControl.ItemTemplate>
				<DataTemplate>
					<!--uc means name space/folder "UserControls"-->
					<uc:TileGroupUC ThisTileGroup="{Binding}" />
				</DataTemplate>
			</ItemsControl.ItemTemplate>
		</ItemsControl>
	</ScrollViewer>
```
and Tile List:
```xml
			<!--Tile List-->
			<ItemsControl Name="TileItemControl">
				<ItemsControl.ItemsPanel>
					<ItemsPanelTemplate>
						<WrapPanel />
					</ItemsPanelTemplate>
				</ItemsControl.ItemsPanel>
				<ItemsControl.ItemTemplate>
					<DataTemplate>
						<!--uc means name space/folder "UserControls"-->
						<uc:TileUC ThisTile="{Binding}" />
					</DataTemplate>
				</ItemsControl.ItemTemplate>
			</ItemsControl>
```

**Comment by maxkatz6** [comment]
> Any idea or solution for this performance issue? Nested list is a common need for most apps.

Profiling via dotTrace or similar would help here.

**Comment by timunie** [comment]
I'd also look into writing my own virtualized Panel here if I expect more than a few items here. As this really helps to improve performance.

**Reply by thevortexcloud** [reply]
There was this virtualised wrap panel that was proposed awhile back. It's probably a good starting point:

https://github.com/AvaloniaUI/Avalonia/pull/11251

**Comment by butaixianran** [comment]
I'll look into creating my own virtualized panel. But I'm still wondering why this happened. It is not that slow at the beginning, it becomes slower and slower. 

Anyway, thanks for your answers.

**Reply by timunie** [reply]
maybe some events are not unregistered?

**Reply by butaixianran** [reply]
Indeed, I'll take care of those events and see if things gonna be better. Thanks.

Edit:
Unreistered all events from those controls not used any more. Nothing better, same result.

**Comment by butaixianran** [comment]
I take 3 days to convert this project into a Win UI 3 project. Which works very well, there is no such issue at all. 

So I won't take more time on this issue. Closed.
